### DISCUSSION TITLE
Data not converted for certain SpecificCharacterSet values or encoding

### DISCUSSION THREAD

**Discussion by quintonn** [dcm4che/dcm4che#941] [Q&A] [answered]
I have a dicom file with Scandinavian characters in it 
> e.g. Åkerblad^Ämtö

The **SpecificCharacterSet** is set to "ISO_IR 100".  
When I receive this dicom and convert it to XML I get the correct output.  
Althoug this only works when I explicitly set the output encoding to UTF-8:  
I.E.  
`transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");`

But when the **SpecificCharacterSet** is set to "ISO_IR 192" the output looks like this:  
> ?kerblad^?mt?^  

I have tried setting the output encoding to all sorts of other values, but none seem to work.  
I have tried using the dcm2xml tool using the "+U8" option and that is able to produce the correct results.  
I think this is not your dcm2xml tool, but the one from dvtk (it's in my command line so not sure).  

Attached is a zip file containing my dicom file:
[sample.zip](https://github.com/dcm4che/dcm4che/files/6325154/sample.zip) 

The code I am using to test this is as follows:
```   
   public void test() throws Exception {
		final DicomInputStream xt = new DicomInputStream(new File("....../files/sample.dcm"));
		final Attributes ab = xt.readDataset();
		System.out.println(ab.getString(Tag.ReferringPhysicianName));

		final InputStream inStream = getClass().getResourceAsStream("files/sample.dcm");
		final byte[] messageBody = IOUtils.toByteArray(inStream);

		final Attributes attributes = convertDicomToAttributes(new ByteArrayInputStream(messageBody), false);

		final String xml = CreateXML(attributes);

		final String pat = attributes.getString(Tag.ReferringPhysicianName);
		System.out.println(pat);

		final boolean containsCorrectValue = xml.contains("Åkerblad");
		//final String tmp = "Åkerblad^Ämtö^Örthén^^Övergaard";
		//System.out.println(tmp);
		System.out.println(xml);
		System.out.println(containsCorrectValue + "");
		Assert.assertTrue(containsCorrectValue);
	}

	private String CreateXML(final Attributes data) throws TransformerConfigurationException, IOException {
		final SAXTransformerFactory tf = (SAXTransformerFactory) TransformerFactory.newInstance();
		final TransformerHandler handler = tf.newTransformerHandler();
		final Transformer transformer = handler.getTransformer();

		transformer.setOutputProperty(OutputKeys.INDENT, "yes");
		//transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");

		final ByteArrayOutputStream b = new ByteArrayOutputStream();
		final DicomOutputStream out = new DicomOutputStream(b, "");
		data.writeTo(out);

		final DicomInputStream in = new DicomInputStream(new ByteArrayInputStream(b.toByteArray()));

		final ByteArrayOutputStream xml = new ByteArrayOutputStream();
		handler.setResult(new StreamResult(xml));

		final SAXWriter saxWriter = new SAXWriter(handler);


		in.setDicomInputHandler(saxWriter);
		in.readDataset(-1, -1);

		in.close();

		return xml.toString("UTF-8");
	}

	private Attributes convertDicomToAttributes(final InputStream dicom, final boolean includeBulkData) throws IOException {
		DicomInputStream dis = null;
		final Path tempDirPath = Files.createTempDirectory("");
		try {
			dis = new DicomInputStream(dicom);
			if (includeBulkData == true) {
				dis.setIncludeBulkData(IncludeBulkData.URI);
			}

			dis.setBulkDataDirectory(tempDirPath.toFile());
			return dis.readDataset(-1, -1);

		} finally {
			SafeClose.close(dis);

			if (includeBulkData == false) {
				try {
					FileUtils.deleteDirectory(tempDirPath.toFile());
				} catch (final IOException e) {
					e.printStackTrace();
				}
			}

		}
	}
```  

Is it possible to convert the dicom to XML and get the output correct?  
Do I need to do something different?  
Any help is appreciated.

**Comment by gunterze** [comment, marked as answer]
The actual encoding of string values in the attached DICOM file is not UTF-8 but Latin 1: e.g.:
```
$ xxd -g1 /tmp/sample.dcm | grep PN | head -n1
000000d0: 90 00 50 4e 20 00 c5 6b 65 72 62 6c 61 64 5e c4  ..PN ..kerblad^.
```

c5 is not a valid UTF-8 code, but the Latin 1 code for Å 


The value of _Specific Character Set (0008,0005)_ has to reflect the actual used Character Set Code of encoded string attributes in the data set. For changing its value from `ISO_IR 100` to `ISO_IR 192` you have to also trans-code all string attributes from Latin 1 to UTF-8!

**Reply by quintonn** [reply]
Hi,
Thanks for the info.

But I am unclear how the **SpecificCharacterSet** affects the actual data we get once converted to XML.  
As I mentioned, if this value is set to **ISO_IR 100** I get the correct characters even when I specify UTF-8 in the encoding.  
The exact same dicom file with SpecificCharacterSet set to **ISO_IR 192** produces different results.  
The only thing that changes is the SpecificCharacterSet.  

I am testing using the DVTK RIS emulator, and all I do between tests is change that 1 value.  

So it's confusing to me that the dicom file's encoding impacts the results.  

At what point in the code I have above does the SpecificCharacterSet value get used by the dcm4che library??

**Reply by gunterze** [reply]
Again, the problem is not the conversion to XML, but that the value of SpecificCharacterSet must describe the actual encoding of string values in the dataset.

Or in other words, you will get the right result also with SpecificCharacterSet ISO_IR 192. if string values in the binary DICOM file are actual encoded in UTF-8.
